<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>snRNA_scRNA_Pipeline Introduction &mdash; snRNAseq scRNAseq Pipeline 0.1 documentation</title>
      <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/tooltipster.custom.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/tooltipster.bundle.min.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/tooltipster-sideTip-shadow.min.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/tooltipster-sideTip-punk.min.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/tooltipster-sideTip-noir.min.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/tooltipster-sideTip-light.min.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/tooltipster-sideTip-borderless.min.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/micromodal.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/sphinx_rtd_theme.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/tables.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/page_style.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/sphinx_highlight.js"></script>
        <script src="_static/js/hoverxref.js"></script>
        <script src="_static/js/tooltipster.bundle.min.js"></script>
        <script src="_static/js/micromodal.min.js"></script>
        <script async="async" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
        <script src="https://unpkg.com/mermaid/dist/mermaid.min.js"></script>
        <script>mermaid.initialize({startOnLoad:true});</script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Salient Features" href="highlights.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="#" class="icon icon-home"> snRNAseq scRNAseq Pipeline
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Introduction</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="highlights.html">Salient Features</a></li>
<li class="toctree-l1"><a class="reference internal" href="set_up.html">Executing the Pipeline</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Getting Started</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="basic_run.html">Understanding Snakemake workflows</a></li>
<li class="toctree-l1"><a class="reference internal" href="basic_run.html#a-basic-pipeline">A basic pipeline</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Understanding the Pipeline</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="sub_workflows.html">Modules</a></li>
<li class="toctree-l1"><a class="reference internal" href="sub_workflows.html#sub-snakemake-workflows">Sub-Snakemake workflows</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Advanced Tutorials</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="complex_tutorials.html">Demultiplex pooled snRNA seq datasets</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Reference</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="glossary.html">Glossary</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="#">snRNAseq scRNAseq Pipeline</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="#" class="icon icon-home"></a></li>
      <li class="breadcrumb-item active">snRNA_scRNA_Pipeline Introduction</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/index.md.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>This documentation is incomplete and is under heavy development!</p>
</div>
<section id="snrna-scrna-pipeline-introduction">
<h1>snRNA_scRNA_Pipeline Introduction<a class="headerlink" href="#snrna-scrna-pipeline-introduction" title="Permalink to this heading"></a></h1>
<section id="todo">
<h2>TODO:<a class="headerlink" href="#todo" title="Permalink to this heading"></a></h2>
<ul class="contains-task-list simple">
<li><p>Add rule-specific <a class="reference external" href="https://snakemake.readthedocs.io/en/stable/snakefiles/deployment.html#integrated-package-management">conda envs</a> and verify them. Conda envs can’t be used with <strong>run</strong>, which is present in the following rules:</p>
<ul class="contains-task-list">
<li class="task-list-item"><p><input class="task-list-item-checkbox" checked="checked" disabled="disabled" type="checkbox"> STARsolo</p></li>
<li class="task-list-item"><p><input class="task-list-item-checkbox" disabled="disabled" type="checkbox"> split_bams</p></li>
<li class="task-list-item"><p><input class="task-list-item-checkbox" disabled="disabled" type="checkbox"> split_bams_gt</p></li>
<li class="task-list-item"><p><input class="task-list-item-checkbox" disabled="disabled" type="checkbox"> kite</p></li>
<li class="task-list-item"><p><input class="task-list-item-checkbox" disabled="disabled" type="checkbox"> pheno_demux3</p></li>
</ul>
</li>
<li class="task-list-item"><p><input class="task-list-item-checkbox" disabled="disabled" type="checkbox"> For rules that use <strong>genefull_matrices</strong> make input function that take either <em>Gene</em> or <em>GeneFull</em> dependent on the project.</p></li>
<li class="task-list-item"><p><input class="task-list-item-checkbox" disabled="disabled" type="checkbox"> Change input for the rules create_inp_splitBams and create_inp_splitBams_gt_demux. Input function that should remove low mito cells.</p></li>
<li class="task-list-item"><p><input class="task-list-item-checkbox" disabled="disabled" type="checkbox"> Beautify the function get_filt_barcodes (pheno_demux3.snkmk).</p></li>
<li class="task-list-item"><p><input class="task-list-item-checkbox" disabled="disabled" type="checkbox"> Employ a strategy for final count matrix dir (file dir in config file) for the cases:</p>
<ul>
<li><p>when both demultiplex software are run simultaneously.</p></li>
<li><p>When there’s an order (try to name each run separately or at least keep the order somewhere mentioned).</p></li>
</ul>
</li>
<li class="task-list-item"><p><input class="task-list-item-checkbox" disabled="disabled" type="checkbox"> Add an option (in config file) to create h5ads when demultiplexing (demultiplex_no_argp.snkmk) or not (can be used as switch when doing gt checks and finalizing donor assignment).</p></li>
<li class="task-list-item"><p><input class="task-list-item-checkbox" disabled="disabled" type="checkbox"> Add an option for the rule cellSNP when ref SNPs vcf need not be subsetted further.</p></li>
<li class="task-list-item"><p><input class="task-list-item-checkbox" disabled="disabled" type="checkbox"> Make the functions similar for demultiplexing with any method.</p></li>
<li class="task-list-item"><p><input class="task-list-item-checkbox" disabled="disabled" type="checkbox"> Move documentation to configargparse.</p></li>
<li class="task-list-item"><p><input class="task-list-item-checkbox" disabled="disabled" type="checkbox"> Write down schemas.</p></li>
<li><p>Add tutorials.</p>
<ul>
<li><p>pooled snRNA seq</p>
<ul class="contains-task-list">
<li class="task-list-item"><p><input class="task-list-item-checkbox" disabled="disabled" type="checkbox"> simple</p></li>
<li class="task-list-item"><p><input class="task-list-item-checkbox" disabled="disabled" type="checkbox"> complex</p></li>
</ul>
</li>
<li><p>scRNA seq</p>
<ul class="contains-task-list">
<li class="task-list-item"><p><input class="task-list-item-checkbox" disabled="disabled" type="checkbox"> simple</p></li>
<li class="task-list-item"><p><input class="task-list-item-checkbox" disabled="disabled" type="checkbox"> complex</p></li>
</ul>
</li>
</ul>
</li>
<li class="task-list-item"><p><input class="task-list-item-checkbox" disabled="disabled" type="checkbox"> Remove dependency on STARsolo as an aligner.</p></li>
<li class="task-list-item"><p><input class="task-list-item-checkbox" disabled="disabled" type="checkbox"> Combine sub-workflows split_bams and split_bams_gt.</p></li>
<li class="task-list-item"><p><input class="task-list-item-checkbox" checked="checked" disabled="disabled" type="checkbox"> Fix issue with reading old wet_lab_info file to update (extension issues).</p></li>
<li class="task-list-item"><p><input class="task-list-item-checkbox" checked="checked" disabled="disabled" type="checkbox"> Some issue with create_wet_lab_info.py file (it misses to add some lines from certain files - try AMP ones)</p></li>
<li class="task-list-item"><p><input class="task-list-item-checkbox" disabled="disabled" type="checkbox"> Add options in config file to allow adding extra params for every software:</p>
<ul class="contains-task-list">
<li class="task-list-item"><p><input class="task-list-item-checkbox" disabled="disabled" type="checkbox"> WASP mode in STARsolo_sort.snkmk</p></li>
</ul>
</li>
<li class="task-list-item"><p><input class="task-list-item-checkbox" disabled="disabled" type="checkbox"> Add new Picard metrics.</p></li>
<li class="task-list-item"><p><input class="task-list-item-checkbox" disabled="disabled" type="checkbox"> Include in new_config.yaml an option to select wasp mode for the rule STARsolo_sort.snkmk</p></li>
<li class="task-list-item"><p><input class="task-list-item-checkbox" disabled="disabled" type="checkbox"> Search Ranking of readthedocs (using config file for this too).</p></li>
<li class="task-list-item"><p><input class="task-list-item-checkbox" disabled="disabled" type="checkbox"> Might incorporate git submodules for repos on git that I use.</p></li>
</ul>
<p>This pipeline intends to not only make complex <a class="hxr-hoverxref hxr-tooltip reference internal" href="glossary.html#term-preprocessing"><span class="xref std std-term">preprocessing</span></a> workflows easy (e.g. snRNA seq with pooled samples, double HTOs, etc.) but also to facilitate the use of common workflows used for preprocessing by providing <em>readymade</em> different combinations of softwares/tools (see <a class="reference internal" href="sub_workflows.html#selectable-modules"><span class="std std-ref">selectable</span></a> modules for more options).</p>
<p>It also supports various software/pipeline for scRNA seq pre-processing.</p>
<p>The highlights of the pipeline are:</p>
<ul>
	<li> Streamlined processes to modify parameters for each program through a single yaml file </li>
	<li> Easily modifiable to accomodate more rules </li>
	<li> Can be used for both individual samples as well as multiplexed pools </li>
	<li> Preserve folder structures (mirroring fastqs' folder structures) </li>
	<li> Organize outputs from each module </li>
	<li> Select multiple pre-set modules that simplifies usage across multiple projects </li>
</ul>
</section>
<section id="changelog">
<h2>Changelog:<a class="headerlink" href="#changelog" title="Permalink to this heading"></a></h2>
<ul class="simple">
<li><p>Changed param name in demultiplex info from <em>Unique genes</em> to <em>gene_ids with an associated gene_name</em>.</p></li>
<li><p>Added new param in demultiplex info file to add more stats when remove gene IDs without an associated gene name.</p></li>
<li><p>Added an option to run cellSNP without any ref vcfs (1000 Genomes Project vcf is min requirement)</p></li>
<li><p>Now create_wet_lab_info scripts can:</p>
<ul>
<li><p>Run without a converter file</p></li>
<li><p>Save donor file along with the wet lab compilation file</p></li>
<li><p>argparse documented</p></li>
</ul>
</li>
<li><p>Fixed an issue with create_wet_lab_info.py file</p></li>
<li><p>create_wet_lab_info.py file now mirrors actions for donor and multiplex compilations.</p></li>
</ul>
</section>
<section id="requirements">
<h2>Requirements<a class="headerlink" href="#requirements" title="Permalink to this heading"></a></h2>
<p>This pipeline depends on the following packages/programs:</p>
<ul>
	<li><a href="https://scanpy.readthedocs.io/en/stable/" name="sc">Scanpy Manual</a></li>
	<li><a href="https://snakemake.readthedocs.io/en/stable/" name="snk">Snakemake Manual</a></li>
	<li><a href="https://github.com/alexdobin/STAR/blob/master/docs/STARsolo.md" name="sts">STARsolo Manual</a> </li>
	<li><a href="https://broadinstitute.github.io/picard/command-line-overview.html#CollectGcBiasMetrics" name="gcb">GC bias metrics (PICARD)</a></li>
	<li><a href="https://broadinstitute.github.io/picard/command-line-overview.html#CollectRnaSeqMetrics" name="rna">RNA seq metrics (PICARD)</a></li>
	<li><a href="https://github.com/pachterlab/kite"name="kite">KITE <i>(kallisto indexing and tag extraction)</i></a></li>
	<li><a href="https://cellsnp-lite.readthedocs.io/en/latest/manual.html" name="csp">cellSNP Manual</a></li>
	<li><a href="https://vireosnp.readthedocs.io/en/latest/manual.html" name="cir">vireoSNP Manual</a></li>
	<li><a href="https://github.com/calico/solo#how-to-demultiplex-cell-hashing-data-using-hashsolo-cli" name="hsolo">hashsolo Info</a></li>
	<li><a href="https://qtltools.github.io/qtltools/pages/QTLtools-mbv.1.html" name="mbv">QTLtools-mbv</a></li>
</ul>
<div class="toctree-wrapper compound">
<p class="caption" role="heading"><span class="caption-text">Introduction</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="highlights.html">Salient Features</a><ul>
<li class="toctree-l2"><a class="reference internal" href="highlights.html#streamlined">Streamlined</a></li>
<li class="toctree-l2"><a class="reference internal" href="highlights.html#maintaining-project-structure">Maintaining project structure</a></li>
<li class="toctree-l2"><a class="reference internal" href="highlights.html#easy-modification-or-addition-of-rules">Easy modification or addition of rules</a></li>
<li class="toctree-l2"><a class="reference internal" href="highlights.html#one-select-running-of-popular-workflows">One-select running of popular workflows</a></li>
<li class="toctree-l2"><a class="reference internal" href="highlights.html#easy-resource-maintenance">Easy resource maintenance</a></li>
<li class="toctree-l2"><a class="reference internal" href="highlights.html#re-attempts-for-known-issues-or-errors">Re-attempts for known issues or errors</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="set_up.html">Executing the Pipeline</a><ul>
<li class="toctree-l2"><a class="reference internal" href="set_up.html#installing-dependencies">Installing dependencies</a></li>
<li class="toctree-l2"><a class="reference internal" href="set_up.html#setting-up-profiles">Setting up profiles</a></li>
<li class="toctree-l2"><a class="reference internal" href="set_up.html#executing-pipeline">Executing Pipeline</a></li>
</ul>
</li>
</ul>
</div>
<div class="toctree-wrapper compound">
<p class="caption" role="heading"><span class="caption-text">Getting Started</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="basic_run.html">Understanding Snakemake workflows</a></li>
<li class="toctree-l1"><a class="reference internal" href="basic_run.html#a-basic-pipeline">A basic pipeline</a><ul>
<li class="toctree-l2"><a class="reference internal" href="basic_run.html#setting-up">Setting up</a></li>
</ul>
</li>
</ul>
</div>
<div class="toctree-wrapper compound">
<p class="caption" role="heading"><span class="caption-text">Understanding the Pipeline</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="sub_workflows.html">Modules</a><ul>
<li class="toctree-l2"><a class="reference internal" href="sub_workflows.html#selectable-modules">Selectable Modules</a></li>
<li class="toctree-l2"><a class="reference internal" href="sub_workflows.html#module-description">Module description</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="sub_workflows.html#sub-snakemake-workflows">Sub-Snakemake workflows</a><ul>
<li class="toctree-l2"><a class="reference internal" href="generated/demul_samples_no_argp.ret_htos_calico_solo.html">demul_samples_no_argp.ret_htos_calico_solo</a></li>
<li class="toctree-l2"><a class="reference internal" href="sub_workflows.html#overview-of-the-pipeline">Overview of the pipeline</a></li>
</ul>
</li>
</ul>
</div>
<div class="toctree-wrapper compound">
<p class="caption" role="heading"><span class="caption-text">Advanced Tutorials</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="complex_tutorials.html">Demultiplex pooled snRNA seq datasets</a><ul>
<li class="toctree-l2"><a class="reference internal" href="complex_tutorials.html#pipeline-overwiew">Pipeline overwiew</a></li>
<li class="toctree-l2"><a class="reference internal" href="complex_tutorials.html#preparing-target-files">Preparing target files</a></li>
<li class="toctree-l2"><a class="reference internal" href="complex_tutorials.html#configuration-file">Configuration File</a></li>
<li class="toctree-l2"><a class="reference internal" href="complex_tutorials.html#project-specific-changes-to-rules">Project-specific changes to rules</a></li>
<li class="toctree-l2"><a class="reference internal" href="complex_tutorials.html#changes-to-executor-script">Changes to executor script</a></li>
</ul>
</li>
</ul>
</div>
<div class="toctree-wrapper compound">
<p class="caption" role="heading"><span class="caption-text">Reference</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="glossary.html">Glossary</a></li>
</ul>
</div>
</section>
</section>
<section id="indices-and-tables">
<h1>Indices and tables<a class="headerlink" href="#indices-and-tables" title="Permalink to this heading"></a></h1>
<ul class="simple">
<li><p><a class="reference internal" href="genindex.html"><span class="std std-ref">Index</span></a></p></li>
<li><p><a class="reference internal" href="py-modindex.html"><span class="std std-ref">Module Index</span></a></p></li>
<li><p><a class="reference internal" href="search.html"><span class="std std-ref">Search Page</span></a></p></li>
</ul>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="highlights.html" class="btn btn-neutral float-right" title="Salient Features" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2022, Prashant N M.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>