<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>scRNA_snRNA_Pipeline &mdash; snRNAseq scRNAseq Pipeline 0.1 documentation</title>
      <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/sphinx_highlight.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="#" class="icon icon-home"> snRNAseq scRNAseq Pipeline
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <!-- Local TOC -->
              <div class="local-toc"><ul>
<li><a class="reference internal" href="#">scRNA_snRNA_Pipeline</a><ul>
<li><a class="reference internal" href="#requirements">Requirements</a><ul>
<li><a class="reference internal" href="#packages-installed-through-conda">Packages installed through conda</a></li>
<li><a class="reference internal" href="#packages-installed-through-pip">Packages installed through pip</a></li>
</ul>
</li>
<li><a class="reference internal" href="#overview-of-the-pipeline">Overview of the pipeline</a></li>
<li><a class="reference internal" href="#setting-up-profiles">Setting up profiles</a></li>
<li><a class="reference internal" href="#executing-pipeline">Executing Pipeline</a></li>
<li><a class="reference internal" href="#configuration-for-pipeline">Configuration for pipeline</a></li>
<li><a class="reference internal" href="#selectable-modules">Selectable Modules</a><ul>
<li><a class="reference internal" href="#module-specifications">Module Specifications</a></li>
</ul>
</li>
<li><a class="reference internal" href="#sub-snakemake-workflows">Sub-Snakemake workflows</a></li>
</ul>
</li>
<li><a class="reference internal" href="#welcome-to-scrnaseq-pipeline-s-documentation">Welcome to scRNAseq Pipeline’s documentation!</a></li>
<li><a class="reference internal" href="#indices-and-tables">Indices and tables</a></li>
</ul>
</div>
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="#">snRNAseq scRNAseq Pipeline</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="#" class="icon icon-home"></a></li>
      <li class="breadcrumb-item active">scRNA_snRNA_Pipeline</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/index.md.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>This documentation is incomplete and is under heavy development!</p>
</div>
<div class="section" id="scrna-snrna-pipeline">
<h1>scRNA_snRNA_Pipeline<a class="headerlink" href="#scrna-snrna-pipeline" title="Permalink to this heading"></a></h1>
<p>Snakemake Pipeline with “selectable” modules for scRNA/snRNA seq pre-processing.
The highlights of the pipeline is:</p>
<ul>
	<li> Streamlined processes to modify parameters for each program through a single yaml file </li>
	<li> Easily modifiable to accomodate more rules </li>
	<li> Can be used for both individual samples as well as multiplexed pools </li>
	<li> Preserve folder structures (mirroring fastqs' folder structures) </li>
	<li> Organize outputs from each module </li>
	<li> Select multiple pre-set modules that simplifies usage across multiple projects </li>
</ul>
<div class="section" id="requirements">
<h2>Requirements<a class="headerlink" href="#requirements" title="Permalink to this heading"></a></h2>
<p>This pipeline depends on the following packages/programs:</p>
<ul>
	<li><a href="https://scanpy.readthedocs.io/en/stable/">Scanpy Manual</a></li>
	<li><a href="https://snakemake.readthedocs.io/en/stable/">Snakemake Manual</a></li>
	<li><a href="https://github.com/alexdobin/STAR/blob/master/docs/STARsolo.md">STARsolo Manual</a> </li>
	<li><a href="https://broadinstitute.github.io/picard/command-line-overview.html#CollectGcBiasMetrics">GC bias metrics (PICARD)</a></li>
	<li><a href="https://broadinstitute.github.io/picard/command-line-overview.html#CollectRnaSeqMetrics">RNA seq metrics (PICARD)</a></li>
	<li><a href="https://github.com/pachterlab/kite">KITE <i>(kallisto indexing and tag extraction)</i></a></li>
	<li><a href="https://cellsnp-lite.readthedocs.io/en/latest/manual.html">cellSNP Manual</a></li>
	<li><a href="https://vireosnp.readthedocs.io/en/latest/manual.html">vireoSNP Manual</a></li>
	<li><a href="https://github.com/calico/solo#how-to-demultiplex-cell-hashing-data-using-hashsolo-cli">hashsolo Info</a></li>
</ul>
<div class="section" id="packages-installed-through-conda">
<h3>Packages installed through conda<a class="headerlink" href="#packages-installed-through-conda" title="Permalink to this heading"></a></h3>
<p>All the packages installed through anaconda3/2018.12 for Python 3.9.5 are described (add_link and file)</p>
</div>
<div class="section" id="packages-installed-through-pip">
<h3>Packages installed through pip<a class="headerlink" href="#packages-installed-through-pip" title="Permalink to this heading"></a></h3>
<p>All the packages installed for R version 4.1.0 are described (add_link and file)</p>
</div>
</div>
<div class="section" id="overview-of-the-pipeline">
<h2>Overview of the pipeline<a class="headerlink" href="#overview-of-the-pipeline" title="Permalink to this heading"></a></h2>
<p>Directed Acyclic Graph of the whole pipeline:
<img alt="DAG" src="_images/Whole_pipeline.png" /></p>
</div>
<div class="section" id="setting-up-profiles">
<h2>Setting up profiles<a class="headerlink" href="#setting-up-profiles" title="Permalink to this heading"></a></h2>
<p>The info for setting up profiles for different workload managers is mentioned <a class="reference external" href="https://github.com/Snakemake-Profiles">here</a></p>
</div>
<div class="section" id="executing-pipeline">
<h2>Executing Pipeline<a class="headerlink" href="#executing-pipeline" title="Permalink to this heading"></a></h2>
<p>This pipeline can be executed by executing (in case of any workflow manager, submitting) the script called <code class="docutils literal notranslate"><span class="pre">run_snakemake.sh</span></code></p>
<div class="highlight-sh notranslate"><div class="highlight"><pre><span></span>sh run_snakemake.sh
</pre></div>
</div>
</div>
<div class="section" id="configuration-for-pipeline">
<h2>Configuration for pipeline<a class="headerlink" href="#configuration-for-pipeline" title="Permalink to this heading"></a></h2>
<p>This pipeline depends on a yaml config file (new_config.yaml), which has all relevant options for each rule present in this pipeline. This not only streamlines the process of maintaining or modifying program specific parameters but also makes repetitive usage (can be within the project or using it over mutliple projects) of this pipeline over time very efficient.</p>
</div>
<div class="section" id="selectable-modules">
<h2>Selectable Modules<a class="headerlink" href="#selectable-modules" title="Permalink to this heading"></a></h2>
<p>The following combinations of programs can be run, where starsolo represents STARsolo, rnaseqmet and gcbiasmet refer to PICARD’s CollectRnaSeqMetrics and CollectGcBiasMetrics, respectively while picard represents inclusion of both programs, kb_solo refers to using kallisto, bustools and calico_solo for demultiplexing, gt_demux refers to using cellSNP and vireoSNP for genotype based demultiplexing, split_bams_kb_solo refers to splitting pooled/multiplexed bams</p>
<ul>
<li> all</li>
<li> starsolo</li>
<li> starsolo_rnaseqmet</li>
<li> starsolo_gcbiasmet</li>
<li> starsolo_kb_solo</li>
<li> starsolo_picard</li>
<li> starsolo_gt_demux</li>
<li> starsolo_split_bams</li>
<li> starsolo_split_bams_gt_demux</li>
<li> starsolo_split_bams_gt_demux_multi_vcf</li>
<li> starsolo_gt_demux_multi_vcf</li>
<li> starsolo_cellsnp</li>
</ul>
<p>NOTE: These module names are case-insensitive!</p>
<div class="section" id="module-specifications">
<h3>Module Specifications<a class="headerlink" href="#module-specifications" title="Permalink to this heading"></a></h3>
<dl>
	<dt>All</dt>
	<dd>This module includes alignment through STARsolo and PICARD's both programs (GcBiasMetrics and RnaSeqMetrics) along with calico solo demultiplexing through kallisto bustools output</dd>
	<dt>STARsolo</dt>
	<dd>Only alignemnt through STARsolo.</dd>
	<dt>STARsolo_rnaseqmet</dt>
	<dd>This module will execute STARsolo and PICARD's RNAseq metrics.</dd>
	<dt>STARsolo_gcbiasmet</dt>
	<dd>This module will execute STARsolo and PICARD's GC bias metrics.</dd>
	<dt>STARsolo_kb_solo</dt>
	<dd>This module will execute STARsolo, kallisto_bustools, and demultiplexing by calico_solo/hashsolo.</dd>
	<dt>STARsolo_PICARD</dt>
	<dd>This module will execute STARsolo, PICARD's RNAseq metrics, and PICARD's GC bias metrics.</dd>
	<dt>STARsolo_gt_demux <i>(Not yet implemented)</i></dt>
	<dd>This module will execute cellSNP and vireoSNP</dd>
</dl>
</div>
</div>
<div class="section" id="sub-snakemake-workflows">
<h2>Sub-Snakemake workflows<a class="headerlink" href="#sub-snakemake-workflows" title="Permalink to this heading"></a></h2>
<p>This pipeline divides each module into its self-contained individual workflows. These are:</p>
<dl>
	<dt> resources.snkmk </dt>
	<dd> </dd>
	<dt> calico_solo_demux.snkmk </dt>
	<dd> </dd>
	<dt> split_bams.snkmk<sup><a href="#ft1" id="ref1" >*</a></sup></dt>
	<dd> </dd>
	<dt> input_processing.snkmk </dt>
	<dd> </dd>
	<dt> STARsolo.snkmk </dt>
	<dd> </dd>
	<dt> produce_targets.snkmk </dt>
	<dd> </dd>
	<dt> snv_aware_align.snkmk (Not yet implemented)</dt>
	<dd> </dd>
	<dt> kite.snkmk </dt>
	<dd> </dd>
	<dt> picard_metrics.snkmk </dt>
	<dd> </dd>
	<dt> pheno_demux3.snkmk </dt>
	<dd> </dd>
	<dt> split_bams_gt.snkmk<sup><a href="#ft1" id="ref1" >*</a></sup></dt>
	<dd> </dd>
	<dt> demultiplex_no_argp.snkmk </dt>
	<dd> </dd>
<p><sup id="#ft1">*</sup> Consolidating into one</p>
</dl>
</div>
</div>
<div class="section" id="welcome-to-scrnaseq-pipeline-s-documentation">
<h1>Welcome to scRNAseq Pipeline’s documentation!<a class="headerlink" href="#welcome-to-scrnaseq-pipeline-s-documentation" title="Permalink to this heading"></a></h1>
<div class="toctree-wrapper compound">
</div>
</div>
<div class="section" id="indices-and-tables">
<h1>Indices and tables<a class="headerlink" href="#indices-and-tables" title="Permalink to this heading"></a></h1>
<ul class="simple">
<li><p><a class="reference internal" href="genindex.html"><span class="std std-ref">Index</span></a></p></li>
<li><p><a class="reference internal" href="py-modindex.html"><span class="std std-ref">Module Index</span></a></p></li>
<li><p><a class="reference internal" href="search.html"><span class="std std-ref">Search Page</span></a></p></li>
</ul>
</div>


           </div>
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2022, Prashant N M.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>