Search.setIndex({"docnames": ["basic_run", "complex_tutorials", "create_Feat_Barc", "create_aggregate_matrix", "create_h5ad_from_bustools", "create_h5ad_from_calico_solo", "create_inp_cellSNP", "create_inp_splitBam", "create_inp_splitBam_gt_demux", "create_wet_lab_info", "demul_samples", "demul_samples_ext_vS", "demultiplex_helper_funcs", "glossary", "highlights", "index", "modules", "set_up", "split_seth5ad_to_samph5ad", "sub_workflows", "update_logs"], "filenames": ["basic_run.md", "complex_tutorials.md", "create_Feat_Barc.rst", "create_aggregate_matrix.rst", "create_h5ad_from_bustools.rst", "create_h5ad_from_calico_solo.rst", "create_inp_cellSNP.rst", "create_inp_splitBam.rst", "create_inp_splitBam_gt_demux.rst", "create_wet_lab_info.rst", "demul_samples.rst", "demul_samples_ext_vS.rst", "demultiplex_helper_funcs.rst", "glossary.md", "highlights.md", "index.md", "modules.rst", "set_up.md", "split_seth5ad_to_samph5ad.rst", "sub_workflows.md", "update_logs.rst"], "titles": ["Understanding Snakemake workflows", "Demultiplex pooled snRNA seq datasets", "create_Feat_Barc module", "create_aggregate_matrix module", "create_h5ad_from_bustools module", "create_h5ad_from_calico_solo module", "create_inp_cellSNP module", "create_inp_splitBam module", "create_inp_splitBam_gt_demux module", "create_wet_lab_info module", "demul_samples module", "demul_samples_ext_vS module", "demultiplex_helper_funcs module", "Glossary", "Salient Features", "snRNA_scRNA_Pipeline Introduction", "helper_py_scripts", "Executing the Pipeline", "split_seth5ad_to_samph5ad module", "Overview of the Pipeline", "update_logs module"], "terms": {"To": [0, 1, 13, 14, 19], "begin": [0, 1], "thi": [0, 1, 2, 4, 5, 6, 7, 12, 13, 14, 15, 17, 19], "setup": [0, 1, 19], "i": [0, 1, 2, 14, 15, 17, 19], "highli": 0, "advis": 0, "go": [0, 14], "through": [0, 1, 15], "": [0, 14, 15, 19], "tutori": [0, 1], "surmis": 0, "easier": 0, "follow": [0, 1, 15, 19], "bottom": 0, "top": 0, "approach": 0, "e": [0, 6, 15, 19], "identifi": [0, 12, 19], "final": [0, 1, 15], "requir": [0, 19], "build": 0, "wai": [0, 1, 15, 19], "infput": 0, "an": [0, 4, 5, 6, 13, 15], "exampl": [0, 13], "dag": [0, 13], "let": 0, "start": [0, 1], "run": [0, 4, 5, 6, 13, 15, 19], "softwar": [0, 7, 14, 15], "preprocess": [0, 13, 15], "scrnaseq": 0, "data": 0, "us": [0, 1, 2, 5, 6, 12, 13, 14, 15, 19], "starsolo": [0, 14, 15, 19], "align": [0, 15], "our": [0, 1], "cdna": 0, "picard": [0, 15, 19], "tool": [0, 6, 15], "while": [0, 19], "retain": [0, 5, 13, 15], "all": [0, 1, 12, 14, 17, 19], "statist": 0, "produc": [0, 5, 12, 13, 19], "For": [0, 13, 14, 15, 19], "case": [0, 13, 15, 17], "we": [0, 1], "2": [0, 1, 7, 12, 19], "from": [0, 1, 4, 5, 12, 13, 15], "output": [0, 4, 5, 7, 12, 14, 15, 19], "program": [0, 14, 15, 19], "collectgcbiasmetr": [0, 19], "collectrnaseqmetr": [0, 19], "thei": 0, "independ": 0, "each": [0, 1, 7, 12, 13, 15, 19], "other": 0, "henc": 0, "were": 0, "onli": [0, 1, 15], "one": [0, 1, 5, 7, 12, 13, 14, 19], "won": 0, "t": 0, "assimil": 0, "read": [0, 1, 15], "runn": 0, "anoth": [0, 12, 14], "script": [0, 2, 5, 6, 15, 17, 19], "run_update_log": 0, "sh": [0, 17], "click": [0, 13], "here": [0, 1, 13, 14, 17], "know": 0, "firstli": [0, 1], "list": [0, 1, 6, 12, 15], "input": [0, 6, 15, 19], "check": [0, 15, 19], "differ": [0, 15, 17, 19], "style": 0, "text": [0, 7], "line": [0, 15], "per": [0, 12, 19], "sampl": [0, 1, 2, 13, 15], "pic": 0, "show": [0, 1], "present": [0, 1, 12, 14, 19], "directori": [0, 19], "content": 0, "pool": [0, 12, 13, 15, 19], "As": [0, 19], "can": [0, 6, 13, 15, 17, 19], "see": [0, 15, 19], "contain": [0, 1, 2, 5, 6, 12, 14, 15, 19], "represent": 0, "doesn": 0, "separ": [0, 1, 12, 15], "r1": 0, "r2": 0, "tree_struct": 0, "complex": [1, 15], "workflow": [1, 15, 17], "simplifi": [1, 15], "streamlin": [1, 15], "make": [1, 12, 15, 19], "more": [1, 5, 14, 15, 19], "interest": 1, "annot": [1, 5, 7], "individu": [1, 13, 15, 19], "genotyp": [1, 19], "base": [1, 19], "cellsnp": [1, 6, 15, 19], "vireosnp": [1, 12, 15, 19], "well": [1, 15], "hto": [1, 2, 12, 15], "kite": [1, 15, 19], "hashsolo": [1, 5, 12, 15, 19], "The": [1, 12, 14, 15, 17, 19], "visual": 1, "need": [1, 12, 15, 19], "creat": [1, 2, 4, 5, 6, 7, 8, 15, 19], "deriv": 1, "which": [1, 13, 14], "input_process": [1, 15, 19], "add": [1, 15], "link": 1, "wildcard": [1, 14, 15], "asdasd": 1, "ani": [1, 15, 17], "utilis": 1, "set": [1, 15, 19], "up": [1, 15, 19], "new_config": [1, 14, 15, 19], "yaml": [1, 14, 15, 19], "config": [1, 15], "ha": [1, 19], "relev": 1, "option": [1, 12, 14, 15, 19], "furthermor": 1, "been": [1, 6], "section": 1, "comment": 1, "sub": [1, 15], "ocur": 1, "order": [1, 15], "appear": 1, "typic": 1, "ar": [1, 12, 14, 15, 17, 19], "certain": [1, 15], "irrespect": 1, "being": 1, "pictur": 1, "showcas": 1, "part": 1, "1": [1, 13, 17, 19], "3": [1, 17, 19], "last_step": 1, "kei": 1, "fed": 1, "pre": [1, 14, 15], "select": [1, 15], "have": [1, 5, 6, 7, 12], "snakefil": [1, 19], "featur": [2, 15], "barcod": [2, 5, 6, 7, 12, 13], "file": [2, 4, 5, 7, 12, 13, 14, 15, 17, 19], "kiteseq": 2, "featurebarcod": 2, "csv": 2, "given": [2, 5], "name": [2, 12, 15, 19], "its": [2, 7, 12, 19], "correspond": [2, 7], "respect": [2, 19], "get_argument_pars": [2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 16, 18, 20], "gener": [2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 18, 20], "return": [2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 18, 20], "argument": [2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 18, 20], "parser": [2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 18, 20], "main": [2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 16, 18, 20], "function": [2, 6, 7, 9, 12, 15, 19], "get_info": [3, 16], "fn": [3, 20], "ser1": [3, 9], "f": 3, "none": [3, 12, 18], "get_names_anno": [3, 16], "gid": 3, "col": [3, 12], "entri": [3, 7, 10, 20], "point": [3, 10, 20], "process_h5ad": [3, 16], "fname": [3, 11, 12, 18], "gen": 3, "grch38": 3, "h5ad": [4, 5, 6, 15], "bustool": [4, 5, 19], "sciprt": 4, "take": [4, 6, 15], "after": 4, "count": [4, 5, 6, 12, 13, 15, 19], "hash": [5, 13], "usual": 5, "obtain": 5, "sofwatr": 5, "like": 5, "matrix": [5, 6, 12, 13, 15, 19], "gene": [5, 12, 13, 15], "often": 5, "than": 5, "mai": 5, "lesser": 5, "them": [5, 7], "compar": 5, "thu": 5, "should": [5, 6], "expect": 5, "max": 5, "number": [5, 12], "hase": 5, "effect": 6, "depend": [6, 15, 19], "condit": [6, 15], "some": [6, 15], "cell": [6, 7, 12, 13, 15], "remov": [6, 15, 19], "g": [6, 15], "alreadi": 6, "demultiplex": [6, 7, 12, 15, 19], "wa": 6, "column": [6, 7, 12], "classif": [6, 12], "classifi": [6, 12], "doublet": [6, 12, 13], "neg": [6, 12], "too": [6, 15], "donor": [7, 12, 13, 15], "first": [7, 12], "second": [7, 12], "If": [7, 12], "multipl": [7, 15, 19], "want": 7, "sep": [7, 12], "save_df": [7, 8, 16], "df": [7, 8], "suff": [7, 8], "op": [7, 8], "get_subid": [9, 16], "df1": 9, "df2": 9, "get_don_id": [11, 12, 16], "x": [11, 12], "t_df": [11, 12], "str": [11, 12], "read_files_ext": [11, 16, 18], "datafram": [11, 12, 18], "ret_subj_id": [11, 12, 16], "ser": [11, 12], "set_don_id": [11, 12, 16], "auto_read": [12, 16], "kwarg": [12, 20], "demux_by_calico_solo": [12, 16], "bc": 12, "seri": 12, "df_": 12, "samp": 12, "col_list": 12, "dem_c": 12, "donor_convert": 12, "bool": 12, "panda": 12, "core": 12, "calico_solo": [12, 15, 19], "assign": [12, 15], "ret_htos_calico_solo": [12, 16], "demux_by_vireo": [12, 16], "vir_out_fil": 12, "conv_df": 12, "vireo": [12, 15, 19], "A": [12, 13, 15], "pd": 12, "path": 12, "donor_id": 12, "tsv": 12, "convert": [12, 15], "chang": [12, 15, 19], "demux": [12, 15], "stat": [12, 15], "demux_stat": [12, 16], "demux_freq": 12, "demux_nam": 12, "covert": 12, "paramet": [12, 15, 19], "string": 12, "repres": [12, 19], "type": 12, "parse_fil": [12, 15, 16], "wet_lab_df": 12, "s_name": 12, "h": 12, "d_con": 12, "union": 12, "int": 12, "inform": [12, 15], "id": [12, 15], "wet": [12, 15], "lab": [12, 15], "subid": 12, "row": 12, "otherwis": 12, "val": 12, "index": [12, 15], "solo": 12, "method": [12, 15], "extra": [12, 14, 15], "convent": 12, "convet": 12, "similar": [12, 14, 15], "also": [12, 15], "suit": 12, "beautifi": [12, 15], "so": 12, "feasibl": 12, "where": [13, 19], "oligo": 13, "tag": [13, 15], "antibodi": 13, "against": 13, "ubiquit": 13, "express": 13, "surfac": 13, "protein": 13, "uniqu": [13, 15], "label": 13, "distinct": 13, "subsequ": 13, "szhl": 13, "18": 13, "singl": [13, 15], "experi": [13, 19], "directli": 13, "analysi": 13, "pipelin": [13, 14], "mean": 13, "either": [13, 15, 19], "attribut": 13, "origin": 13, "direct": [13, 15], "acycl": 13, "graph": 13, "marlon": 13, "stoeckiu": 13, "shiwei": 13, "zheng": 13, "brian": 13, "houck": 13, "loomi": 13, "stephani": 13, "hao": 13, "bertrand": 13, "z": 13, "yeung": 13, "william": 13, "m": 13, "mauck": 13, "3rd": 13, "peter": 13, "smibert": 13, "rahul": 13, "satija": 13, "enabl": 13, "multiplex": [13, 15, 19], "detect": 13, "genom": [13, 15], "biol": 13, "19": 13, "224": 13, "decemb": 13, "2018": [13, 17], "heart": 14, "modifi": [14, 15, 19], "properti": 14, "place": 14, "info": [14, 15, 17, 19], "promin": 14, "abil": 14, "captur": 14, "folder": [14, 15, 19], "fastq": [14, 15], "provid": [14, 15, 19], "organ": [14, 15], "variou": [14, 15], "vari": 14, "manner": 14, "understand": 14, "ad": [14, 15], "param": [14, 15], "add_her": 14, "mani": [14, 19], "document": 15, "incomplet": 15, "under": 15, "heavi": 15, "develop": 15, "miscellan": 15, "write": 15, "down": 15, "schema": 15, "snrna": 15, "seq": 15, "scrna": 15, "doubl": 15, "rule": [15, 19], "genefull_matric": 15, "geneful": 15, "project": [15, 19], "combin": [15, 19], "split_bam": [15, 19], "split_bams_gt": [15, 19], "search": 15, "rank": 15, "readthedoc": 15, "might": [15, 19], "incorpor": 15, "git": 15, "submodul": 15, "repo": 15, "new": 15, "metric": [15, 19], "allow": 15, "everi": 15, "rerun": 15, "those": 15, "updat": 15, "log": 15, "fix": 15, "demultiplex_helper_func": [15, 16], "py": 15, "analyse_vireo": 15, "snakemake_rul": 15, "calico_solo_demux": [15, 19], "emploi": 15, "strategi": 15, "dir": 15, "when": 15, "both": [15, 19], "simultan": 15, "try": 15, "least": 15, "keep": 15, "somewher": 15, "mention": [15, 17, 19], "includ": 15, "append": 15, "helper_funct": 15, "identify_swap": [15, 19], "pheno_demux3": [15, 19], "get_filt_barcod": 15, "picard_metr": [15, 19], "produce_target": [15, 19], "target": 15, "resourc": [15, 19], "time": [15, 19], "mem": 15, "low": 15, "mito": 15, "wasp": 15, "mode": 15, "intend": 15, "easi": 15, "etc": 15, "facilit": 15, "common": 15, "readymad": 15, "modul": [15, 16], "It": [15, 19], "support": 15, "process": 15, "highlight": 15, "easili": 15, "accomod": 15, "preserv": 15, "structur": [15, 19], "mirror": 15, "usag": 15, "across": 15, "gene_id": 15, "associ": 15, "gene_nam": 15, "without": 15, "ref": 15, "vcf": [15, 19], "1000": 15, "min": 15, "now": 15, "create_wet_lab_info": [15, 16], "save": 15, "along": 15, "compil": 15, "argpars": 15, "issu": 15, "action": 15, "demux_samples_calico_solo_starsolo": 15, "demux_sampl": 15, "demux_info": 15, "posit": 15, "demultiplex_no_argp": [15, 19], "snkmk": [15, 19], "handl": 15, "work": 15, "demul_samples_no_argp": 15, "demul_sampl": [15, 16], "switch": 15, "do": 15, "gt": 15, "snp": 15, "subset": 15, "further": 15, "old": 15, "wet_lab_info": 15, "extens": 15, "miss": 15, "amp": 15, "ones": 15, "packag": [15, 19], "scanpi": 15, "manual": 15, "snakemak": 15, "gc": 15, "bia": 15, "rna": 15, "kallisto": [15, 19], "extract": 15, "qtltool": [15, 19], "mbv": [15, 19], "salient": 15, "maintain": 15, "modif": 15, "addit": 15, "One": 15, "popular": 15, "mainten": 15, "re": 15, "attempt": 15, "known": 15, "error": 15, "execut": [15, 19], "instal": 15, "profil": 15, "basic": 15, "overview": 15, "descript": 15, "dataset": 15, "overwiew": 15, "prepar": 15, "configur": 15, "specif": [15, 19], "executor": 15, "glossari": 15, "page": 15, "create_feat_barc": 16, "create_aggregate_matrix": 16, "create_h5ad_from_bustool": 16, "create_h5ad_from_calico_solo": 16, "create_inp_cellsnp": 16, "create_inp_splitbam": 16, "create_inp_splitbam_gt_demux": 16, "demul_samples_ext_v": 16, "split_seth5ad_to_samph5ad": 16, "check_sam": [16, 18], "get_donor": [16, 18], "update_log": 16, "calc_ratio": [16, 20], "get_df": [16, 20], "get_filenam": [16, 20], "get_latest_extra_column": [16, 20], "write_log": [16, 20], "anaconda3": 17, "12": 17, "python": [17, 19], "9": 17, "5": 17, "describ": 17, "add_link": 17, "version": 17, "r": [17, 19], "4": 17, "0": 17, "workload": 17, "manag": 17, "submit": 17, "call": 17, "run_snakemak": 17, "ann_d": 18, "files_l": 18, "ann_d2": 18, "boolean": 18, "inp_df": 18, "col_val1": 18, "col_val2": 18, "d": 18, "mh": 18, "sn_f": 18, "sn_g": 18, "is_log": 18, "batch": 18, "dict": 18, "lev": 18, "perform": 19, "how": 19, "accord": 19, "wherev": 19, "possibl": 19, "current": 19, "sure": 19, "global": 19, "variabl": 19, "purpos": 19, "There": 19, "achiev": 19, "aforement": 19, "built": 19, "keyword": 19, "starsolo_rnaseqmet": 19, "starsolo_gcbiasmet": 19, "starsolo_kb_solo": 19, "starsolo_picard": 19, "starsolo_gt_demux": 19, "starsolo_split_bam": 19, "starsolo_split_bams_gt_demux": 19, "starsolo_split_bams_gt_demux_multi_vcf": 19, "starsolo_gt_demux_multi_vcf": 19, "starsolo_cellsnp": 19, "starsolo_rnaseqmet_kb_solo": 19, "starsolo_gcbiasmet_kb_solo": 19, "starsolo_gt_demux_identify_swap": 19, "starsolo_resolve_swaps_gt_demux": 19, "rnaseqmet": 19, "gcbiasmet": 19, "refer": 19, "inclus": 19, "previous": 19, "kb_solo": 19, "gt_demux": 19, "split": 19, "bam": 19, "split_bams_gt_demux": 19, "qtltools_mbv": 19, "multi_vcf": 19, "muiltipl": 19, "same": 19, "worflow": 19, "involv": 19, "module_info": 19, "desc": 19, "own": 19, "sub_wkfl": 19, "all_multi_vcf": 19, "divid": 19, "self": 19, "These": 19, "memori": 19, "mb": 19, "thread": 19, "minut": 19, "custom": 19, "collect": 19, "valu": 19, "snv_aware_align": 19, "soon": 19, "gcbiasmetr": 19, "rnaseqmetr": 19, "swap": 19, "consolid": 19, "Not": 19, "yet": 19, "implement": 19, "numer": 20, "denom": 20, "inp_path": 20, "loc_dir": 20, "file_struct": 20, "suffix": 20, "big_df": 20, "mapper": 20, "all_files_dict": 20, "no_prog": 20}, "objects": {"": [[2, 0, 0, "-", "create_Feat_Barc"], [3, 0, 0, "-", "create_aggregate_matrix"], [4, 0, 0, "-", "create_h5ad_from_bustools"], [5, 0, 0, "-", "create_h5ad_from_calico_solo"], [6, 0, 0, "-", "create_inp_cellSNP"], [7, 0, 0, "-", "create_inp_splitBam"], [8, 0, 0, "-", "create_inp_splitBam_gt_demux"], [9, 0, 0, "-", "create_wet_lab_info"], [10, 0, 0, "-", "demul_samples"], [11, 0, 0, "-", "demul_samples_ext_vS"], [12, 0, 0, "-", "demultiplex_helper_funcs"], [18, 0, 0, "-", "split_seth5ad_to_samph5ad"], [20, 0, 0, "-", "update_logs"]], "create_Feat_Barc": [[2, 1, 1, "", "get_argument_parser"], [2, 1, 1, "", "main"]], "create_aggregate_matrix": [[3, 1, 1, "", "get_argument_parser"], [3, 1, 1, "", "get_info"], [3, 1, 1, "", "get_names_anno"], [3, 1, 1, "", "main"], [3, 1, 1, "", "process_h5ad"]], "create_h5ad_from_bustools": [[4, 1, 1, "", "get_argument_parser"], [4, 1, 1, "", "main"]], "create_h5ad_from_calico_solo": [[5, 1, 1, "", "get_argument_parser"], [5, 1, 1, "", "main"]], "create_inp_cellSNP": [[6, 1, 1, "", "get_argument_parser"], [6, 1, 1, "", "main"]], "create_inp_splitBam": [[7, 1, 1, "", "get_argument_parser"], [7, 1, 1, "", "main"], [7, 1, 1, "", "save_df"]], "create_inp_splitBam_gt_demux": [[8, 1, 1, "", "get_argument_parser"], [8, 1, 1, "", "main"], [8, 1, 1, "", "save_df"]], "create_wet_lab_info": [[9, 1, 1, "", "get_argument_parser"], [9, 1, 1, "", "get_subid"], [9, 1, 1, "", "main"]], "demul_samples": [[10, 1, 1, "", "get_argument_parser"], [10, 1, 1, "", "main"]], "demul_samples_ext_vS": [[11, 1, 1, "", "get_argument_parser"], [11, 1, 1, "", "get_don_ids"], [11, 1, 1, "", "main"], [11, 1, 1, "", "read_files_ext"], [11, 1, 1, "", "ret_subj_ids"], [11, 1, 1, "", "set_don_ids"]], "demultiplex_helper_funcs": [[12, 1, 1, "", "auto_read"], [12, 1, 1, "", "demux_by_calico_solo"], [12, 1, 1, "", "demux_by_vireo"], [12, 1, 1, "", "demux_stats"], [12, 1, 1, "", "get_don_ids"], [12, 1, 1, "", "parse_file"], [12, 1, 1, "", "ret_htos_calico_solo"], [12, 1, 1, "", "ret_subj_ids"], [12, 1, 1, "", "set_don_ids"]], "split_seth5ad_to_samph5ad": [[18, 1, 1, "", "check_same"], [18, 1, 1, "", "get_argument_parser"], [18, 1, 1, "", "get_donors"], [18, 1, 1, "", "main"], [18, 1, 1, "", "read_files_ext"]], "update_logs": [[20, 1, 1, "", "calc_ratio"], [20, 1, 1, "", "get_argument_parser"], [20, 1, 1, "", "get_df"], [20, 1, 1, "", "get_filename"], [20, 1, 1, "", "get_latest_extra_columns"], [20, 1, 1, "", "main"], [20, 1, 1, "", "write_logs"]]}, "objtypes": {"0": "py:module", "1": "py:function"}, "objnames": {"0": ["py", "module", "Python module"], "1": ["py", "function", "Python function"]}, "titleterms": {"understand": [0, 15], "snakemak": [0, 19], "workflow": [0, 14, 19], "A": 0, "basic": 0, "pipelin": [0, 1, 15, 17, 19], "set": [0, 17], "up": [0, 17], "how": 0, "fastq": [0, 1], "file": [0, 1], "ar": 0, "arrang": 0, "creat": 0, "fastq_fil": 0, "txt": 0, "demultiplex": 1, "pool": 1, "snrna": 1, "seq": 1, "dataset": 1, "overwiew": 1, "prepar": 1, "target": 1, "structur": [1, 14], "configur": 1, "common": 1, "project": [1, 14], "specif": 1, "paramet": 1, "dag": 1, "control": 1, "info": 1, "param": 1, "folder": 1, "extra": 1, "can": 1, "remov": 1, "soon": 1, "modul": [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 18, 19, 20], "selector": 1, "chang": 1, "rule": [1, 14], "executor": 1, "script": 1, "create_feat_barc": 2, "create_aggregate_matrix": 3, "create_h5ad_from_bustool": 4, "create_h5ad_from_calico_solo": 5, "create_inp_cellsnp": 6, "create_inp_splitbam": 7, "create_inp_splitbam_gt_demux": 8, "create_wet_lab_info": 9, "demul_sampl": 10, "demul_samples_ext_v": 11, "demultiplex_helper_func": 12, "paramt": 12, "glossari": 13, "salient": 14, "featur": 14, "streamlin": 14, "maintain": 14, "easi": 14, "modif": 14, "addit": 14, "One": 14, "select": [14, 19], "run": 14, "popular": 14, "resourc": 14, "mainten": 14, "re": 14, "attempt": 14, "known": 14, "issu": 14, "error": 14, "snrna_scrna_pipelin": 15, "introduct": 15, "todo": 15, "changelog": 15, "requir": 15, "get": 15, "start": 15, "advanc": 15, "tutori": 15, "refer": 15, "indic": 15, "tabl": 15, "helper_py_script": 16, "execut": 17, "instal": 17, "depend": 17, "packag": 17, "through": 17, "conda": 17, "pip": 17, "profil": 17, "split_seth5ad_to_samph5ad": 18, "overview": 19, "wildcard": 19, "process": 19, "list": 19, "sampl": 19, "descript": 19, "modules_info": 19, "sub": 19, "sub_snakemake_workflows_t": 19, "update_log": 20}, "envversion": {"sphinx.domains.c": 2, "sphinx.domains.changeset": 1, "sphinx.domains.citation": 1, "sphinx.domains.cpp": 8, "sphinx.domains.index": 1, "sphinx.domains.javascript": 2, "sphinx.domains.math": 2, "sphinx.domains.python": 3, "sphinx.domains.rst": 2, "sphinx.domains.std": 2, "sphinx.ext.todo": 2, "sphinxcontrib.bibtex": 9, "sphinx": 57}, "alltitles": {"Demultiplex pooled snRNA seq datasets": [[1, "demultiplex-pooled-snrna-seq-datasets"]], "Pipeline overwiew": [[1, "pipeline-overwiew"]], "Preparing target files": [[1, "preparing-target-files"]], "Fastq File Structure": [[1, "fastq-file-structure"]], "Configuration File": [[1, "configuration-file"]], "Common (project-specific) parameters": [[1, "common-project-specific-parameters"]], "DAG control and project info params": [[1, "dag-control-and-project-info-params"]], "Folder structures": [[1, "folder-structures"]], "Extra Info (can be removed soon!)": [[1, "extra-info-can-be-removed-soon"]], "Module selector": [[1, "module-selector"]], "Project-specific changes to rules": [[1, "project-specific-changes-to-rules"]], "Changes to executor script": [[1, "changes-to-executor-script"]], "Glossary": [[13, "glossary"]], "Salient Features": [[14, "salient-features"]], "Streamlined": [[14, "streamlined"]], "Maintaining project structure": [[14, "maintaining-project-structure"]], "Easy modification or addition of rules": [[14, "easy-modification-or-addition-of-rules"]], "One-select running of popular workflows": [[14, "one-select-running-of-popular-workflows"]], "Easy resource maintenance": [[14, "easy-resource-maintenance"]], "Re-attempts for known issues or errors": [[14, "re-attempts-for-known-issues-or-errors"]], "Executing the Pipeline": [[17, "executing-the-pipeline"]], "Installing dependencies": [[17, "installing-dependencies"]], "Packages installed through conda": [[17, "packages-installed-through-conda"]], "Packages installed through pip": [[17, "packages-installed-through-pip"]], "Setting up profiles": [[17, "setting-up-profiles"]], "Executing Pipeline": [[17, "executing-pipeline"]], "Overview of the Pipeline": [[19, "overview-of-the-pipeline"]], "Wildcard Processing": [[19, "wildcard-processing"]], "List of samples": [[19, "list-of-samples"]], "Selectable Modules": [[19, "selectable-modules"]], "Module description": [[19, "module-description"]], "Modules_info": [[19, "module-info-table"]], "Sub-Snakemake workflows": [[19, "sub-snakemake-workflows"]], "Sub_Snakemake_workflows_table": [[19, "sub-smk-info-table"]], "snRNA_scRNA_Pipeline Introduction": [[15, "snrna-scrna-pipeline-introduction"]], "TODO:": [[15, "todo"]], "Changelog:": [[15, "changelog"]], "Requirements": [[15, "requirements"]], "Introduction": [[15, null]], "Getting Started": [[15, null]], "Understanding the Pipeline": [[15, null]], "Advanced Tutorials": [[15, null]], "Reference": [[15, null]], "Indices and tables": [[15, "indices-and-tables"]], "Understanding Snakemake workflows": [[0, "understanding-snakemake-workflows"]], "A basic pipeline": [[0, "a-basic-pipeline"]], "Setting up": [[0, "setting-up"]], "How fastq files are arranged": [[0, "how-fastq-files-are-arranged"]], "Create fastq_files.txt": [[0, "create-fastq-files-txt"]], "create_Feat_Barc module": [[2, "module-create_Feat_Barc"]], "create_h5ad_from_bustools module": [[4, "module-create_h5ad_from_bustools"]], "create_h5ad_from_calico_solo module": [[5, "module-create_h5ad_from_calico_solo"]], "create_inp_cellSNP module": [[6, "module-create_inp_cellSNP"]], "create_inp_splitBam module": [[7, "module-create_inp_splitBam"]], "create_inp_splitBam_gt_demux module": [[8, "module-create_inp_splitBam_gt_demux"]], "create_wet_lab_info module": [[9, "module-create_wet_lab_info"]], "demul_samples module": [[10, "module-demul_samples"]], "demul_samples_ext_vS module": [[11, "module-demul_samples_ext_vS"]], "demultiplex_helper_funcs module": [[12, "module-demultiplex_helper_funcs"]], "Paramters": [[12, "paramters"]], "helper_py_scripts": [[16, "helper-py-scripts"]], "split_seth5ad_to_samph5ad module": [[18, "module-split_seth5ad_to_samph5ad"]], "update_logs module": [[20, "module-update_logs"]], "create_aggregate_matrix module": [[3, "module-create_aggregate_matrix"]]}, "indexentries": {"create_feat_barc": [[2, "module-create_Feat_Barc"]], "get_argument_parser() (in module create_feat_barc)": [[2, "create_Feat_Barc.get_argument_parser"]], "main() (in module create_feat_barc)": [[2, "create_Feat_Barc.main"]], "module": [[2, "module-create_Feat_Barc"], [3, "module-create_aggregate_matrix"], [4, "module-create_h5ad_from_bustools"], [5, "module-create_h5ad_from_calico_solo"], [6, "module-create_inp_cellSNP"], [7, "module-create_inp_splitBam"], [8, "module-create_inp_splitBam_gt_demux"], [9, "module-create_wet_lab_info"], [10, "module-demul_samples"], [11, "module-demul_samples_ext_vS"], [12, "module-demultiplex_helper_funcs"], [18, "module-split_seth5ad_to_samph5ad"], [20, "module-update_logs"]], "create_aggregate_matrix": [[3, "module-create_aggregate_matrix"]], "get_argument_parser() (in module create_aggregate_matrix)": [[3, "create_aggregate_matrix.get_argument_parser"]], "get_info() (in module create_aggregate_matrix)": [[3, "create_aggregate_matrix.get_info"]], "get_names_anno() (in module create_aggregate_matrix)": [[3, "create_aggregate_matrix.get_names_anno"]], "main() (in module create_aggregate_matrix)": [[3, "create_aggregate_matrix.main"]], "process_h5ad() (in module create_aggregate_matrix)": [[3, "create_aggregate_matrix.process_h5ad"]], "create_h5ad_from_bustools": [[4, "module-create_h5ad_from_bustools"]], "get_argument_parser() (in module create_h5ad_from_bustools)": [[4, "create_h5ad_from_bustools.get_argument_parser"]], "main() (in module create_h5ad_from_bustools)": [[4, "create_h5ad_from_bustools.main"]], "create_h5ad_from_calico_solo": [[5, "module-create_h5ad_from_calico_solo"]], "get_argument_parser() (in module create_h5ad_from_calico_solo)": [[5, "create_h5ad_from_calico_solo.get_argument_parser"]], "main() (in module create_h5ad_from_calico_solo)": [[5, "create_h5ad_from_calico_solo.main"]], "create_inp_cellsnp": [[6, "module-create_inp_cellSNP"]], "get_argument_parser() (in module create_inp_cellsnp)": [[6, "create_inp_cellSNP.get_argument_parser"]], "main() (in module create_inp_cellsnp)": [[6, "create_inp_cellSNP.main"]], "create_inp_splitbam": [[7, "module-create_inp_splitBam"]], "get_argument_parser() (in module create_inp_splitbam)": [[7, "create_inp_splitBam.get_argument_parser"]], "main() (in module create_inp_splitbam)": [[7, "create_inp_splitBam.main"]], "save_df() (in module create_inp_splitbam)": [[7, "create_inp_splitBam.save_df"]], "create_inp_splitbam_gt_demux": [[8, "module-create_inp_splitBam_gt_demux"]], "get_argument_parser() (in module create_inp_splitbam_gt_demux)": [[8, "create_inp_splitBam_gt_demux.get_argument_parser"]], "main() (in module create_inp_splitbam_gt_demux)": [[8, "create_inp_splitBam_gt_demux.main"]], "save_df() (in module create_inp_splitbam_gt_demux)": [[8, "create_inp_splitBam_gt_demux.save_df"]], "create_wet_lab_info": [[9, "module-create_wet_lab_info"]], "get_argument_parser() (in module create_wet_lab_info)": [[9, "create_wet_lab_info.get_argument_parser"]], "get_subid() (in module create_wet_lab_info)": [[9, "create_wet_lab_info.get_subid"]], "main() (in module create_wet_lab_info)": [[9, "create_wet_lab_info.main"]], "demul_samples": [[10, "module-demul_samples"]], "get_argument_parser() (in module demul_samples)": [[10, "demul_samples.get_argument_parser"]], "main() (in module demul_samples)": [[10, "demul_samples.main"]], "demul_samples_ext_vs": [[11, "module-demul_samples_ext_vS"]], "get_argument_parser() (in module demul_samples_ext_vs)": [[11, "demul_samples_ext_vS.get_argument_parser"]], "get_don_ids() (in module demul_samples_ext_vs)": [[11, "demul_samples_ext_vS.get_don_ids"]], "main() (in module demul_samples_ext_vs)": [[11, "demul_samples_ext_vS.main"]], "read_files_ext() (in module demul_samples_ext_vs)": [[11, "demul_samples_ext_vS.read_files_ext"]], "ret_subj_ids() (in module demul_samples_ext_vs)": [[11, "demul_samples_ext_vS.ret_subj_ids"]], "set_don_ids() (in module demul_samples_ext_vs)": [[11, "demul_samples_ext_vS.set_don_ids"]], "auto_read() (in module demultiplex_helper_funcs)": [[12, "demultiplex_helper_funcs.auto_read"]], "demultiplex_helper_funcs": [[12, "module-demultiplex_helper_funcs"]], "demux_by_calico_solo() (in module demultiplex_helper_funcs)": [[12, "demultiplex_helper_funcs.demux_by_calico_solo"]], "demux_by_vireo() (in module demultiplex_helper_funcs)": [[12, "demultiplex_helper_funcs.demux_by_vireo"]], "demux_stats() (in module demultiplex_helper_funcs)": [[12, "demultiplex_helper_funcs.demux_stats"]], "get_don_ids() (in module demultiplex_helper_funcs)": [[12, "demultiplex_helper_funcs.get_don_ids"]], "parse_file() (in module demultiplex_helper_funcs)": [[12, "demultiplex_helper_funcs.parse_file"]], "ret_htos_calico_solo() (in module demultiplex_helper_funcs)": [[12, "demultiplex_helper_funcs.ret_htos_calico_solo"]], "ret_subj_ids() (in module demultiplex_helper_funcs)": [[12, "demultiplex_helper_funcs.ret_subj_ids"]], "set_don_ids() (in module demultiplex_helper_funcs)": [[12, "demultiplex_helper_funcs.set_don_ids"]], "dag": [[13, "term-DAG"]], "pooled sample": [[13, "term-pooled-sample"]], "preprocessing": [[13, "term-preprocessing"]], "check_same() (in module split_seth5ad_to_samph5ad)": [[18, "split_seth5ad_to_samph5ad.check_same"]], "get_argument_parser() (in module split_seth5ad_to_samph5ad)": [[18, "split_seth5ad_to_samph5ad.get_argument_parser"]], "get_donors() (in module split_seth5ad_to_samph5ad)": [[18, "split_seth5ad_to_samph5ad.get_donors"]], "main() (in module split_seth5ad_to_samph5ad)": [[18, "split_seth5ad_to_samph5ad.main"]], "read_files_ext() (in module split_seth5ad_to_samph5ad)": [[18, "split_seth5ad_to_samph5ad.read_files_ext"]], "split_seth5ad_to_samph5ad": [[18, "module-split_seth5ad_to_samph5ad"]], "calc_ratio() (in module update_logs)": [[20, "update_logs.calc_ratio"]], "get_argument_parser() (in module update_logs)": [[20, "update_logs.get_argument_parser"]], "get_df() (in module update_logs)": [[20, "update_logs.get_df"]], "get_filename() (in module update_logs)": [[20, "update_logs.get_filename"]], "get_latest_extra_columns() (in module update_logs)": [[20, "update_logs.get_latest_extra_columns"]], "main() (in module update_logs)": [[20, "update_logs.main"]], "update_logs": [[20, "module-update_logs"]], "write_logs() (in module update_logs)": [[20, "update_logs.write_logs"]]}})