<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Modules &mdash; snRNAseq scRNAseq Pipeline 0.1 documentation</title>
      <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/tooltipster.custom.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/tooltipster.bundle.min.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/tooltipster-sideTip-shadow.min.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/tooltipster-sideTip-punk.min.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/tooltipster-sideTip-noir.min.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/tooltipster-sideTip-light.min.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/tooltipster-sideTip-borderless.min.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/micromodal.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/sphinx_rtd_theme.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/tables.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/page_style.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/sphinx_highlight.js"></script>
        <script src="_static/js/hoverxref.js"></script>
        <script src="_static/js/tooltipster.bundle.min.js"></script>
        <script src="_static/js/micromodal.min.js"></script>
        <script async="async" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
        <script src="https://unpkg.com/mermaid/dist/mermaid.min.js"></script>
        <script>mermaid.initialize({startOnLoad:true});</script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="helper_py_scripts" href="modules.html" />
    <link rel="prev" title="Understanding Snakemake workflows" href="basic_run.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="index.html" class="icon icon-home"> snRNAseq scRNAseq Pipeline
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Introduction</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="highlights.html">Salient Features</a></li>
<li class="toctree-l1"><a class="reference internal" href="set_up.html">Executing the Pipeline</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Getting Started</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="basic_run.html">Understanding Snakemake workflows</a></li>
<li class="toctree-l1"><a class="reference internal" href="basic_run.html#a-basic-pipeline">A basic pipeline</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Understanding the Pipeline</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="current reference internal" href="#">Modules</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#selectable-modules">Selectable Modules</a></li>
<li class="toctree-l2"><a class="reference internal" href="#module-description">Module description</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="#sub-snakemake-workflows">Sub-Snakemake workflows</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#overview-of-the-pipeline">Overview of the pipeline</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="modules.html">helper_py_scripts</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Advanced Tutorials</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="complex_tutorials.html">Demultiplex pooled snRNA seq datasets</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Reference</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="glossary.html">Glossary</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">snRNAseq scRNAseq Pipeline</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home"></a></li>
      <li class="breadcrumb-item active">Modules</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/sub_workflows.md.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="modules">
<h1>Modules<a class="headerlink" href="#modules" title="Permalink to this heading"></a></h1>
<p>This pipeline has many combinations of the aforementioned programs as a built-in set that can be executed using specific keywords.</p>
<section id="selectable-modules">
<span id="id1"></span><h2>Selectable Modules<a class="headerlink" href="#selectable-modules" title="Permalink to this heading"></a></h2>
<p>The following combinations of programs can be run:</p>
<ul>
<li> all</li>
<li> starsolo</li>
<li> starsolo_rnaseqmet</li>
<li> starsolo_gcbiasmet</li>
<li> starsolo_kb_solo</li>
<li> starsolo_picard</li>
<li> starsolo_gt_demux</li>
<li> starsolo_split_bams</li>
<li> starsolo_split_bams_gt_demux</li>
<li> starsolo_split_bams_gt_demux_multi_vcf</li>
<li> starsolo_gt_demux_multi_vcf</li>
<li> starsolo_cellsnp</li>
<li>starsolo_rnaseqmet_kb_solo</li>
<li>starsolo_gcbiasmet_kb_solo</li>
<li>starsolo_gt_demux_identify_swaps</li>
<li>starsolo_resolve_swaps_gt_demux <sup><small>#</small></sup></li>
</ul>
<p>where <strong>starsolo</strong> represents STARsolo; <strong>rnaseqmet</strong> and <strong>gcbiasmet</strong> refer to PICARD’s CollectRnaSeqMetrics and CollectGcBiasMetrics, respectively while <strong>picard</strong> represents inclusion of both the previously-mentioned programs; <strong>kb_solo</strong> refers to using kallisto, bustools and calico_solo for demultiplexing; <strong>gt_demux</strong> refers to using cellSNP and vireoSNP for genotype based demultiplexing; <strong>split_bams</strong> refers to splitting pooled/multiplexed bams using hashsolo’s outputs while <strong>split_bams_gt_demux</strong> refers to splitting pooled/multiplexed bams using vireo’s output; <strong>identify_swaps</strong> refers to using qtltools_mbv. The option <strong>multi_vcf</strong> is to provide muiltiple runs (i.e. multiple sets of vcf inputs) for the same sample.</p>
</section>
<section id="module-description">
<h2>Module description<a class="headerlink" href="#module-description" title="Permalink to this heading"></a></h2>
<table class="docutils align-default" id="module-info-table">
<caption><span class="caption-text">Modules_info</span><a class="headerlink" href="#module-info-table" title="Permalink to this table"></a></caption>
<colgroup>
<col style="width: 33%" />
<col style="width: 33%" />
<col style="width: 33%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Module Name</p></th>
<th class="head"><p>Module Info</p></th>
<th class="head"><p>Sub Worflows Involved</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>all</p></td>
<td><p>module_info (more desc in its own file)</p></td>
<td><p>sub_wkfl</p></td>
</tr>
<tr class="row-odd"><td><p>starsolo</p></td>
<td><p>module_info (more desc in its own file)</p></td>
<td><p>sub_wkfl</p></td>
</tr>
<tr class="row-even"><td><p>starsolo_rnaseqmet</p></td>
<td><p>module_info (more desc in its own file)</p></td>
<td><p>sub_wkfl</p></td>
</tr>
<tr class="row-odd"><td><p>starsolo_gcbiasmet</p></td>
<td><p>module_info (more desc in its own file)</p></td>
<td><p>sub_wkfl</p></td>
</tr>
<tr class="row-even"><td><p>starsolo_kb_solo</p></td>
<td><p>module_info (more desc in its own file)</p></td>
<td><p>sub_wkfl</p></td>
</tr>
<tr class="row-odd"><td><p>starsolo_picard</p></td>
<td><p>module_info (more desc in its own file)</p></td>
<td><p>sub_wkfl</p></td>
</tr>
<tr class="row-even"><td><p>starsolo_gt_demux</p></td>
<td><p>module_info (more desc in its own file)</p></td>
<td><p>sub_wkfl</p></td>
</tr>
<tr class="row-odd"><td><p>starsolo_split_bams</p></td>
<td><p>module_info (more desc in its own file)</p></td>
<td><p>sub_wkfl</p></td>
</tr>
<tr class="row-even"><td><p>starsolo_split_bams_gt_demux</p></td>
<td><p>module_info (more desc in its own file)</p></td>
<td><p>sub_wkfl</p></td>
</tr>
<tr class="row-odd"><td><p>starsolo_split_bams_gt_demux_multi_vcf</p></td>
<td><p>module_info (more desc in its own file)</p></td>
<td><p>sub_wkfl</p></td>
</tr>
<tr class="row-even"><td><p>starsolo_gt_demux_multi_vcf</p></td>
<td><p>module_info (more desc in its own file)</p></td>
<td><p>sub_wkfl</p></td>
</tr>
<tr class="row-odd"><td><p>starsolo_cellsnp</p></td>
<td><p>module_info (more desc in its own file)</p></td>
<td><p>sub_wkfl</p></td>
</tr>
<tr class="row-even"><td><p>starsolo_rnaseqmet_kb_solo</p></td>
<td><p>module_info (more desc in its own file)</p></td>
<td><p>sub_wkfl</p></td>
</tr>
<tr class="row-odd"><td><p>starsolo_gcbiasmet_kb_solo</p></td>
<td><p>module_info (more desc in its own file)</p></td>
<td><p>sub_wkfl</p></td>
</tr>
<tr class="row-even"><td><p>starsolo_gt_demux_identify_swaps</p></td>
<td><p>module_info (more desc in its own file)</p></td>
<td><p>sub_wkfl</p></td>
</tr>
<tr class="row-odd"><td><p>starsolo_resolve_swaps_gt_demux</p></td>
<td><p>module_info (more desc in its own file)</p></td>
<td><p>sub_wkfl</p></td>
</tr>
</tbody>
</table>
</section>
</section>
<section id="sub-snakemake-workflows">
<h1>Sub-Snakemake workflows<a class="headerlink" href="#sub-snakemake-workflows" title="Permalink to this heading"></a></h1>
<p>This pipeline divides each module into its self-contained individual workflows. These are:</p>
<table class="docutils align-default" id="sub-smk-info-table">
<caption><span class="caption-text">Sub_Snakemake_workflows_table</span><a class="headerlink" href="#sub-smk-info-table" title="Permalink to this table"></a></caption>
<colgroup>
<col style="width: 50%" />
<col style="width: 50%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Name of Workflow</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>resources.snkmk</p></td>
<td><p>It contains memory (in MB per thread) and time requirements (in minutes) for each rule.</p></td>
</tr>
<tr class="row-odd"><td><p>calico_solo_demux.snkmk</p></td>
<td><p>It contains hashsolo rule.</p></td>
</tr>
<tr class="row-even"><td><p>split_bams.snkmk <a class="footnote-reference brackets" href="#consol" id="id2">1</a></p></td>
<td><p>It contains rules needed to split pooled bams into individual bams dependent on output
produced by either hashsolo or vireoSNP using custom scripts.</p></td>
</tr>
<tr class="row-odd"><td><p>input_processing.snkmk</p></td>
<td><p>It contains rules that collects values for all the wildcards.</p></td>
</tr>
<tr class="row-even"><td><p>STARsolo.snkmk</p></td>
<td><p>It contains rules for STARsolo.</p></td>
</tr>
<tr class="row-odd"><td><p>produce_targets.snkmk</p></td>
<td><p>It contains the <strong>rule all</strong> and the needed functions.</p></td>
</tr>
<tr class="row-even"><td><p>snv_aware_align.snkmk <a class="footnote-reference brackets" href="#not-yet" id="id3">2</a></p></td>
<td><p><strong>This might be removed soon</strong></p></td>
</tr>
<tr class="row-odd"><td><p>kite.snkmk</p></td>
<td><p>It contains rules for the <a class="reference external" href="https://github.com/pachterlab/kite#kite-kallisto-indexing-and-tag-extraction">kite</a> workflow.</p></td>
</tr>
<tr class="row-even"><td><p>picard_metrics.snkmk</p></td>
<td><p>It contains rules for all PICARD metrics (GCBiasMetrics and RNAseqMetrics).</p></td>
</tr>
<tr class="row-odd"><td><p>pheno_demux3.snkmk</p></td>
<td><p>It contains rules for the cellSNP-vireoSNP pipeline.</p></td>
</tr>
<tr class="row-even"><td><p>split_bams_gt.snkmk <a class="footnote-reference brackets" href="#consol" id="id4">1</a></p></td>
<td><p>It contains rules needed to split pooled bams into individual bams dependent on output produced by vireo.</p></td>
</tr>
<tr class="row-odd"><td><p>demultiplex_no_argp.snkmk</p></td>
<td><p>It contains rules for demultiplexing using hashsolo and/or vireoSNP output and create a count matrix file.</p></td>
</tr>
<tr class="row-even"><td><p>identify_swaps.snkmk</p></td>
<td><p>It contains rules for identifying swaps using QTLtools-mbv.</p></td>
</tr>
</tbody>
</table>
<table class="autosummary longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">demul_samples_no_argp.ret_htos_calico_solo</span></code></p></td>
<td><p>Return HTO information and classification for each cell barcode.</p></td>
</tr>
</tbody>
</table>
<section id="overview-of-the-pipeline">
<h2>Overview of the pipeline<a class="headerlink" href="#overview-of-the-pipeline" title="Permalink to this heading"></a></h2>
<p>Directed Acyclic Graph of the whole pipeline:
<img alt="DAG" src="_images/Whole_pipeline.png" /></p>
<hr class="footnotes docutils" />
<dl class="footnote brackets">
<dt class="label" id="consol"><span class="brackets">1</span><span class="fn-backref">(<a href="#id2">1</a>,<a href="#id4">2</a>)</span></dt>
<dd><p>Consolidating into one</p>
</dd>
<dt class="label" id="not-yet"><span class="brackets"><a class="fn-backref" href="#id3">2</a></span></dt>
<dd><p>Not yet implemented</p>
</dd>
</dl>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="basic_run.html" class="btn btn-neutral float-left" title="Understanding Snakemake workflows" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="modules.html" class="btn btn-neutral float-right" title="helper_py_scripts" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2022, Prashant N M.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>